src/gtest/test_rpc_wallet.cpp

```diff
@@ -1,1258 +1,1274 @@
+        EXPECT_NO_THROW(CallRPC(string("z_importkey ") + testSaplingKey));
+        EXPECT_NO_THROW(retValue = CallRPC(string("z_exportkey ") + testSaplingAddr));
+        EXPECT_EQ(retValue.get_str(), testSaplingKey);
+    }
+
+    // Verify we can list the keys imported
+    EXPECT_NO_THROW(retValue = CallRPC("z_listaddresses"));
+    UniValue arr = retValue.get_array();
+    EXPECT_EQ(arr.size() , n1);
+
+    // Put addresses into a set
+    unordered_set<string> myaddrs;
+    for (UniValue element : arr.getValues()) {
+        myaddrs.insert(element.get_str());
+    }
+
+    // Verify number of addresses stored in wallet is n1
+    size_t numAddrs = myaddrs.size();
+    EXPECT_EQ(numAddrs , n1);
+    pwalletMain->GetSaplingPaymentAddresses(saplingAddrs);
+    EXPECT_EQ(saplingAddrs.size() , numAddrs);
+
+    // Ask wallet to list addresses
+    EXPECT_NO_THROW(retValue = CallRPC("z_listaddresses"));
+    arr = retValue.get_array();
+    EXPECT_EQ(arr.size() , numAddrs);
+
+    // Create a set from them
+    unordered_set<string> listaddrs;
+    for (UniValue element : arr.getValues()) {
+        listaddrs.insert(element.get_str());
+    }
+
+    // Verify the two sets of addresses are the same
+    EXPECT_EQ(listaddrs.size() , numAddrs);
+    EXPECT_EQ(myaddrs , listaddrs);
+
+     // Check if too many args
+    EXPECT_THROW(CallRPC("z_getnewaddress toomanyargs"), runtime_error);
+}
+
+// Check if address is of given type and spendable from our wallet.
+template <typename ADDR_TYPE>
+void CheckHaveAddr(const libzcash::PaymentAddress& addr) {
+
+    EXPECT_TRUE(IsValidPaymentAddress(addr));
+    auto addr_of_type = get_if<ADDR_TYPE>(&addr);
+    BOOST_ASSERT(addr_of_type != nullptr);
+
+    HaveSpendingKeyForPaymentAddress test(pwalletMain);
+    EXPECT_TRUE(test(*addr_of_type));
+}
+
+TEST_F(TestRpcWallet, rpc_wallet_z_getnewaddress)
+{
+    using namespace libzcash;
+    UniValue addr;
+
+    if (!pwalletMain->HaveHDSeed()) {
+        pwalletMain->GenerateNewSeed();
+    }
+
+    KeyIO keyIO(Params());
+    // No parameter defaults to sapling address
+    addr = CallRPC("z_getnewaddress");
+    CheckHaveAddr<SaplingPaymentAddress>(keyIO.DecodePaymentAddress(addr.get_str()));
+
+    // Passing 'sapling' should also work
+    addr = CallRPC("z_getnewaddress sapling");
+    CheckHaveAddr<SaplingPaymentAddress>(keyIO.DecodePaymentAddress(addr.get_str()));
+
+    // Should throw on invalid argument
+    CheckRPCThrows("z_getnewaddress garbage", "Invalid address type");
+
+    // Too many arguments will throw with the help
+    EXPECT_THROW(CallRPC("z_getnewaddress many args"), runtime_error);
+}
+
+/**
+ * Test Async RPC operations.
+ * Tip: Create mock operations by subclassing AsyncRPCOperation.
+ */
+
+class MockSleepOperation : public AsyncRPCOperation {
+public:
+    chrono::milliseconds naptime;
+    MockSleepOperation(int t=1000) {
+        this->naptime = chrono::milliseconds(t);
+    }
+    virtual ~MockSleepOperation() {
```

Comment:
remove it - default virtual destructor is declared in the parent class AsyncRPCOperation (BTW with noexcept)